src/lib.rs

```diff
@@ -1,55 +1,97 @@
-//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Consensus] struct
-//! requires access to an [Environment] object which black-boxes the network layer and gives
-//! appropriate access to the set of available blocks that we need to make consensus on.
+//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Member] struct
+//! requires access to a network layer, a cryptographic primitive, and a data provided that
+//! gives appropriate access to the set of available data that we need to make consensus on.
 
 use codec::{Decode, Encode};
-use futures::{Future, Sink, Stream};
-use log::{debug, error};
+use futures::Future;
 use std::{
     fmt::{Debug, Display},
-    hash::Hash,
-};
-use tokio::{
-    sync::{mpsc, oneshot},
-    time::Duration,
+    hash::Hash as StdHash,
 };
+use tokio::sync::mpsc;
 
-use crate::{creator::Creator, extender::Extender, syncer::Syncer, terminal::Terminal};
+use crate::nodes::{NodeCount, NodeIndex, NodeMap};
 
-pub use crate::nodes::{NodeCount, NodeIndex, NodeMap};
+pub use member::{Config, Member};
 
+mod bft;
+mod consensus;
 mod creator;
 mod extender;
+mod member;
 pub mod nodes;
 mod syncer;
 mod terminal;
 mod testing;
+mod units;
+
+pub trait DataProvider<Data> {
+    type Error: Debug;
+    fn get_data(&self) -> Data;
+    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
+}
+
+pub trait KeyBox<Signature>: MyIndex {
+    fn sign(&self, msg: &Vec<u8>) -> Signature;
+    fn verify(&self, msg: &Vec<u8>, sgn: &Signature) -> bool;
+}
+
+#[async_trait::async_trait]
+pub trait Network {
+    type Error: Debug;
+    fn send(&self, command: NetworkCommand) -> Result<(), Self::Error>;
+    async fn recv(&mut self) -> Option<NetworkEvent>;
+}
+
+#[derive(Clone, Debug, Encode, Decode)]
+pub enum NetworkCommand {
+    SendToAll(Vec<u8>),
+    SendToPeer(Vec<u8>, Vec<u8>),
+    SendToRandPeer(Vec<u8>),
+    ReliableBroadcast(Vec<u8>),
+}
+
+#[derive(Clone, Debug)]
+pub enum NetworkEvent {
+    MessageReceived(Vec<u8>, Vec<u8>),
+}
+
+pub type SessionId = u64;
 
 pub trait MyIndex {
```

Comment:
`MyIndex` sounds like a name for a trait in some tutorial. The instances of this types are not indices... maybe rename to `HasIndex` or `IndexOwner`?

Reply:
`Indexed`? Or even `Index` I guess, this would fit the naming scheme in Rust in general.

Reply:
Yeah, I agree the name is not great. I would propose MemberIndex.

Reply:
I've changed to `Index` as this not always is used for Member.